idr/README.md

```diff
@@ -0,0 +1,97 @@
+# IDR
+
+**IDR** == **I**ntermediate **D**ata **R**epresentation or **I**n-memory **D**ata **R**epresentation
+
+IDR is an intermediate data structure used by omniparser ingesters to store raw data read from various
+formats of inputs, including CSV/txt/XML/EDI/JSON/etc, and then used by schema handlers to perform
+transforms. It is flexible and versatile to represent all kinds of data formats supported (or to be
+supported) by omniparser.
+
+*Credit:* The basic data structures and various operations and algorithms used by IDR are mostly
+inherited/adapted from, modified based on, and inspired by works done in https://github.com/antchfx/xmlquery
+and https://github.com/antchfx/xpath. Thank you very much!
+
+The basic building block of an IDR is a `Node` and an IDR is in fact a `Node` tree. Each `Node` has
+two parts (see actual code [here](./node.go)):
+```
+type Node struct {
+	Parent, FirstChild, LastChild, PrevSibling, NextSibling *Node
+
+	Type NodeType
```

Comment:
Will `NodeType` be generic across all different formats, or it is already mixed different ones? E.g. regarding json input format, we can't have `AttributeNode`?
Not sure whether it should belong to format specific or not.

Reply:
Nothing in the world can be 100% pure, 1) `Type` is generic enough 2) 3 out of 4 type values needed across all formats, so it doesn't make sense to tease that apart. Another goal is not carrying extra fields when not needed. The  old design has `Prefix`, `NamespaceURI` and `Attr` in addition to `Type` and `Data`, ALL the time. But right now, for all other formats, the node just carries, `Type`/`Data` and mostly and empty `FormatSpecific` (except JSON) field. A lot cleaner.